change `zarr_read_kwargs` to `mode` argument

melonora opened the pull request:
- [x] Closes #12204 
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

With #12153 I forgot to update one warning. Based on issue #12204 I went back to the code state before this PR to check whether `zarr_read_kwargs` is indeed the correct argument to use from now on. The problem was that `**kwargs` was not only used for `zarr.create` as stated in the docstring back then. It was also use when creating a `store` from a url. This could involve both `read_only` and `mode` arguments. Having `read_only` does not make sense to me when writing so in this PR I added an argument `mode` and remove `zarr_read_kwargs`. 

I also added some checks for `mode`. Mode `r` will throw an error. If `mode=="w"` then the `overwrite` argument for `zarr.create_array` or `zarr.create` will be set to `True`. Interestingly with a sharded array and if an array is already stored at the location, this will cause a checksum error when `lock==False`. I am not certain yet whether this is a bug in zarr. For now `lock==True` when `mode=="w"` prevents the issue.

@jacobtomlinson @CSSFrancis @thewtex @jo-mueller

melonora commented:
That would indeed not be required. It could be unpacked with that name though just for more clarity and I wiuld also update the docstring with the actual documentation links. I do not have access to my pc at the moment, but will do so later

melonora commented:
That would indeed not be required. It could be unpacked with that name though just for more clarity and I wiuld also update the docstring with the actual documentation links. I do not have access to my pc at the moment, but will do so later

melonora commented:
@CSSFrancis changes have been performed

melonora commented:
FYI @will-moore

melonora commented:
@TomAugspurger Do you think this PR is good to go so it could be included in the first release of 2026?

TomAugspurger commented:
Let's get this or https://github.com/dask/dask/pull/12233 in today (cc @joshua-gould).

Anybody mind summarizing the differences? My main goal is to ensure that code from dask<=2025.11.0 continues to work. If we can also make code from dask==2025.12.0 then great, but IMO that's secondary.

melonora commented:
@TomAugspurger if you would like we could have a quick minicall (15 min would be enough). 

But in short yes, in the othere PR the read arguments get mixed with the actual `Group.create_array` arguments. In this PR given that the only valid argument is `mode` I just added that parameter so now `**kwargs` are really only arguments for `create_array`. Also in the other PR if you pass `mode` in `**kwargs` it will have no effect because it is not passed on where it should be, which is `_setup_zarr_store(url, storage_options, **zarr_read_kwargs)`.

However, this PR is not fully backward compatible as far as I am aware if somebody was passing on `read_only` in `**kwargs`. The reason why I removed it is because I don't know why somebody would make it `read_only` when writing. The `mode` argument I kept because there are multiple modes allowing for writing.

melonora commented:
@TomAugspurger I will try to be there with the next community meeting if that is more suitable. I saw that 2026.1.0 got released, but I think it would be good to get this in sooner rather than later given the API change. We don't want people to adopt the current API too much if it gets switched back to `**kwargs`.

melonora commented:
@TomAugspurger Done, can you please let me know if it is clear to you?

melonora commented:
> > Thanks @melonora.
> > @joshua-gould are you able to check whether this fixes the issue you ran into?
> 
> This fixes my issue, thanks.

Great!
